BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) 2020 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.biometricloginsample
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.text.Editable
+import android.text.TextWatcher
+import android.util.Log
+import android.view.inputmethod.EditorInfo
+import androidx.appcompat.app.AppCompatActivity
+import androidx.appcompat.widget.AppCompatButton
+import androidx.appcompat.widget.AppCompatEditText
+import androidx.appcompat.widget.AppCompatTextView
+import androidx.biometric.BiometricManager
+import androidx.biometric.BiometricPrompt
+import androidx.lifecycle.Observer
+import androidx.lifecycle.ViewModelProviders
+
+
+/**
+ * 1) after entering "valid" username and password, login button becomes enabled
+ * 2) User clicks biometrics?
+ *   - a) if no template exists, then ask user to register template
+ *   - b) if template exists, ask user to confirm by entering username & password
+ */
+class MainActivity : AppCompatActivity() {
+    val TAG = "MainActivity"
+
+    private lateinit var loginWithPasswordViewModel: LoginWithPasswordViewModel
+
+    private lateinit var biometricPrompt: BiometricPrompt
+    private lateinit var cryptographyManager: CryptographyManager
+    private lateinit var encryptedServerTokenWrapper: CiphertextWrapper
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        Log.d(TAG, "onCreate is called")
+
+    }
+
+    override fun onResume() {
+        super.onResume()
+        Log.d(TAG, "onResume called")
+        cryptographyManager = CryptographyManager()
```

Comment:
This can be created in the constructor, removing the need for it to be `lateinit`.

Reply:
Please elaborate

Reply:
Basically, it's typical to setup member variables in `onCreate`, since that's called only once, but `onResume` is called each time the app comes to the foreground, which will recreate `CryptographyManager` each time.

Then, because the constructor for `CryptographyManager` doesn't take any parameters, it should be created above.